librosa/core/spectrum.py

```diff
@@ -1281,6 +1282,152 @@ def fmt(y, t_min=0.5, n_fmt=None, kind='cubic', beta=0.5, over_sample=1, axis=-1
     return result[idx] * np.sqrt(n) / n_fmt
 
 
+@cache(level=30)
+def pcen(S, sr=22050, hop_length=512, gain=0.98, bias=2, power=0.5,
+         time_constant=0.395, eps=1e-6, b=None, max_size=1):
+    '''Per-channel energy normalization (PCEN) [1]_
+
+    This function normalizes a time-frequency representation `S` by
+    performing automatic gain control, followed by non-linear compression:
+
+        P = (S / (eps + M)**gain + bias)**power - bias**power
+
+    where `M` is the low-pass filtered output of `S`:
+
+        M[f, t] = (1 - b) M[f, t - 1] + b * S[f, t]
+
+    and if `b` is not provided, it is calculated as:
+
+        b = 1 - exp(-hop_length / (sr * time_constant))
+
+    This normalization is designed to suppress background noise, and
+    emphasize foreground signals, and can be used as an alternative to
+    decibel scaling (`amplitude_to_db`).
+
+    This implementation also supports frequency-bin smoothing by specifying
+    `max_size > 1`.  If this option is used, the filtered spectrogram `M` is
+    computed as
+
+        M[f, t] = (1 - b) M[f, t - 1] + b * R[f, t]
+        
+    where `R` has been max-filtered along the frequency axis, similar to
+    the Superflux algorithm implemented in `onset.onset_strength`:
+
+        R[f, t] = max(R[f - max_size//2: f + max_size//2, t])
+
+    This can be used to perform automatic gain control on signals that cross
+    or span multiple frequency bans, which may be desirable for spectrograms
+    with high frequency resolution.
+
+    .. [1] Wang, Y., Getreuer, P., Hughes, T., Lyon, R. F., & Saurous, R. A.
+       (2017, March). Trainable frontend for robust and far-field keyword spotting.
+       In Acoustics, Speech and Signal Processing (ICASSP), 2017
+       IEEE International Conference on (pp. 5670-5674). IEEE.
+
+    Parameters
+    ----------
+    S : np.ndarray (non-negative) [shape=(n, m)]
+        input spectrogram
+
+    sr : number > 0 [scalar]
+        Sampling rate of audio
+
+    hop_length : int > 0 [scalar]
+        Hop length of `S`
+
+    gain : number > 0 [scalar]
+        Gain factor.  Typical values should be slightly less than 1.
+
+    bias : number >= 0 [scalar]
+        The bias point of the non-linear compression
+
+    power : number > 0 [scalar]
+        The compression factor.  Typical values should be between 0 and 1.
```

Comment:
It's not a factor, it's an exponent.
Maybe say that the lower the `power`, the stronger the compression?